database/rrdcalc.c

```diff
@@ -662,6 +662,7 @@ static void rrdcalc_labels_unlink_alarm_loop(RRDHOST *host, RRDCALC *alarms) {
 }
 
 void rrdcalc_labels_unlink_alarm_from_host(RRDHOST *host) {
+    rrdhost_check_rdlock(host);
```

Comment:
I don't understand why this line has been added, please walk me through why it is here. As far as I understand the `rdhost_check_rdlock()` call, it checks that a read-lock is currently held on the `rwlock` and it aborts Netdata if this is not true (through the `fatal()` call). So we are using this as a run-time assertion that there is a read-lock held on the `rwlock`. 
Looking at the call-site for this procedure in health.c:157 there is a `write-lock` acquired on the host immediately above the call. This does not seem to match the assumption that it is read-locked?

Reply:
We need at least a read lock and we don't mind to have a write lock. The `rdhost_check_rdlock()` call uses `pthread_rwlock_trywrlock()` underneath, so it basically means "any lock" is already acquired.

Reply:
Yes I agree about where the code goes. When I looked up the manpage I was working from this description:
> The pthread_rwlock_wrlock() function shall apply a write lock to the read-write lock referenced by rwlock. The calling thread acquires the write lock if no other thread (reader or writer) holds the read-write lock rwlock. Otherwise, the thread shall block until it can acquire the lock. The calling thread may deadlock if at the time the call is made it holds the read-write lock (whether a read or write lock).

So, are we relying on `EDEADLK` to indicate that the calling thread aleady holds the lock? I could not see anywhere in the code that distinguished this case from `EBUSY`. What have I missed?

Reply:
We can see the number of the error in the debug message:
```
int ret = __netdata_rwlock_trywrlock(rwlock);

debug(D_LOCKS, "RW_LOCK: netdata_rwlock_trywrlock(0x%p) = %d in %llu usec, from %lu@%s, %s()", rwlock, ret, ... );
```

Reply:
Don't forget, that `rdhost_check_rdlock()` is for debugging only.